File: apps/frontend/src/services/entities.ts
import api from './api'
import type { EntitySchema, EntitySchemaCreate, EntitySchemaUpdate } from '../types/schema'

// Backend response type (snake_case)
interface EntityResponse {
  id: string
  project_id: string
  name: string
  description: string | null
  version: number
  schema: any
  rules_config: string | null
  rules_format: string | null
  created_at: string
  updated_at: string
}

// Transform backend response to frontend EntitySchema
function transformEntityResponse(data: EntityResponse): EntitySchema {
  return {
    id: data.id,
    name: data.name,
    description: data.description || undefined,
    version: data.version,
    tables: data.schema.tables || [],
    relationships: data.schema.relationships,
    layout: data.schema.layout,
    updatedAt: data.updated_at,
    rulesConfig: data.rules_config || undefined,
    rulesFormat: (data.rules_format as 'yaml' | 'json') || undefined,
  }
}

export async function autosaveEntity(projectId: string | undefined, entity: EntitySchema) {
  // Save to localStorage first (immediate, never fails)
  try {
    const key = `autosave:${projectId || 'default'}:${entity.id}`
    localStorage.setItem(key, JSON.stringify(entity))
  } catch (e) {
    // noop
  }
  
  // Then save to backend database (best effort)
  if (projectId) {
    try {
      const schema = {
        tables: entity.tables || [],
        relationships: entity.relationships,
        layout: entity.layout
      }
      await api.put(`/api/v1/projects/${projectId}/entities/${entity.id}`, {
        name: entity.name,
        description: entity.description,
        schema
      })
    } catch (error: any) {
      // If entity doesn't exist, create it
      if (error?.response?.status === 404) {
        try {
          const response = await api.post(`/api/v1/projects/${projectId}/entities`, {
            id: entity.id,
            name: entity.name,
            description: entity.description,
            schema: {
              tables: entity.tables || [],
              relationships: entity.relationships,
              layout: entity.layout
            }
          })
          
          // If backend created with different ID, update localStorage to match
          const createdId = response.data?.id
          if (createdId && createdId !== entity.id) {
            const oldKey = `autosave:${projectId}:${entity.id}`
            const newKey = `autosave:${projectId}:${createdId}`
            const data = localStorage.getItem(oldKey)
            if (data) {
              const entityData = JSON.parse(data)
              entityData.id = createdId
              localStorage.setItem(newKey, JSON.stringify(entityData))
              localStorage.removeItem(oldKey)
            }
          }
        } catch (e) {
          console.warn('Backend entity creation failed:', e)
        }
      } else {
        console.warn('Backend entity save failed:', error)
      }
    }
  }
}

export async function listEntities(projectId: string) {
  const res = await api.get(`/api/v1/projects/${projectId}/entities`)
  return (res.data as EntityResponse[]).map(transformEntityResponse)
}

export async function getEntity(projectId: string, entityId: string) {
  const res = await api.get(`/api/v1/projects/${projectId}/entities/${entityId}`)
  return transformEntityResponse(res.data as EntityResponse)
}

export async function createEntity(projectId: string, payload: EntitySchemaCreate) {
  const res = await api.post(`/api/v1/projects/${projectId}/entities`, payload)
  return transformEntityResponse(res.data as EntityResponse)
}

export async function updateEntity(projectId: string, entityId: string, patch: EntitySchemaUpdate) {
  const res = await api.put(`/api/v1/projects/${projectId}/entities/${entityId}`, patch)
  return transformEntityResponse(res.data as EntityResponse)
}

export async function deleteEntity(projectId: string, entityId: string) {
  await api.delete(`/api/v1/projects/${projectId}/entities/${entityId}`)
  return { id: entityId }
}

File: apps/backend/app/api/api_v1/endpoints/entities.py
"""
API endpoints for wizard entity management.
"""
from typing import List
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import BaseModel, Field

from app.db.session import get_db
from app.db.models import WizardEntity, Project

router = APIRouter()


class EntitySchema(BaseModel):
    """Entity schema for wizard configuration."""
    tables: List[dict] = Field(default_factory=list)
    relationships: List[dict] | None = Field(default=None)
    layout: dict | None = Field(default=None)


class EntityCreate(BaseModel):
    """Create entity request."""
    id: str | None = Field(None, description="Optional UUID for entity. If not provided, one will be generated.")
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(None, description="Optional description of the entity")
    schema: EntitySchema
    rules_config: str | None = Field(None, description="YAML or JSON rules configuration")
    rules_format: str | None = Field(None, description="Format of rules_config: 'yaml' or 'json'")


class EntityUpdate(BaseModel):
    """Update entity request."""
    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = Field(None, description="Optional description of the entity")
    schema: EntitySchema | None = None
    rules_config: str | None = Field(None, description="YAML or JSON rules configuration")
    rules_format: str | None = Field(None, description="Format of rules_config: 'yaml' or 'json'")


class EntityResponse(BaseModel):
    """Entity response."""
    id: str
    project_id: str
    name: str
    description: str | None = None
    version: int
    schema: dict
    rules_config: str | None = None
    rules_format: str | None = None
    created_at: str
    updated_at: str

    class Config:
        from_attributes = True


@router.post("", response_model=EntityResponse, status_code=201)
async def create_entity(
    project_id: UUID,
    entity: EntityCreate,
    db: AsyncSession = Depends(get_db),
):
    """Create a new wizard entity configuration."""
    # Verify project exists
    stmt = select(Project).where(Project.id == project_id)
    result = await db.execute(stmt)
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    # Create entity
    entity_kwargs = {
        "project_id": project_id,
        "name": entity.name,
        "description": entity.description,
        "schema_json": entity.schema.model_dump(),
        "rules_config": entity.rules_config,
        "rules_format": entity.rules_format,
    }
    # Use provided ID if valid UUID string, otherwise let DB generate
    if entity.id:
        try:
            entity_kwargs["id"] = UUID(entity.id)
        except (ValueError, AttributeError):
            pass  # Let DB generate UUID
    
    db_entity = WizardEntity(**entity_kwargs)
    db.add(db_entity)
    await db.commit()
    await db.refresh(db_entity)
    
    return EntityResponse(
        id=str(db_entity.id),
        project_id=str(db_entity.project_id),
        name=db_entity.name,
        description=db_entity.description,
        version=db_entity.version,
        schema=db_entity.schema_json,
        rules_config=db_entity.rules_config,
        rules_format=db_entity.rules_format,
        created_at=db_entity.created_at.isoformat(),
        updated_at=db_entity.updated_at.isoformat(),
    )


@router.get("", response_model=list[EntityResponse])
async def list_entities(
    project_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """List all wizard entities for a project."""
    stmt = (
        select(WizardEntity)
        .where(WizardEntity.project_id == project_id)
        .order_by(WizardEntity.updated_at.desc())
    )
    result = await db.execute(stmt)
    entities = result.scalars().all()
    
    return [
        EntityResponse(
            id=str(e.id),
            project_id=str(e.project_id),
            name=e.name,
            description=e.description,
            version=e.version,
            schema=e.schema_json,
            rules_config=e.rules_config,
            rules_format=e.rules_format,
            created_at=e.created_at.isoformat(),
            updated_at=e.updated_at.isoformat(),
        )
        for e in entities
    ]


@router.get("/{entity_id}", response_model=EntityResponse)
async def get_entity(
    project_id: UUID,
    entity_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Get a specific wizard entity."""
    stmt = select(WizardEntity).where(
        WizardEntity.id == entity_id,
        WizardEntity.project_id == project_id,
    )
    result = await db.execute(stmt)
    entity = result.scalar_one_or_none()
    
    if not entity:
        raise HTTPException(status_code=404, detail="Entity not found")
    
    return EntityResponse(
        id=str(entity.id),
        project_id=str(entity.project_id),
        name=entity.name,
        schema=entity.schema_json,
        created_at=entity.created_at.isoformat(),
        updated_at=entity.updated_at.isoformat(),
    )


@router.put("/{entity_id}", response_model=EntityResponse)
async def update_entity(
    project_id: UUID,
    entity_id: UUID,
    update: EntityUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Update a wizard entity."""
    stmt = select(WizardEntity).where(
        WizardEntity.id == entity_id,
        WizardEntity.project_id == project_id,
    )
    result = await db.execute(stmt)
    entity = result.scalar_one_or_none()
    
    if not entity:
        raise HTTPException(status_code=404, detail="Entity not found")
    
    if update.name is not None:
        entity.name = update.name
    if update.description is not None:
        entity.description = update.description
    if update.schema is not None:
        entity.schema_json = update.schema.model_dump()
    if update.rules_config is not None:
        entity.rules_config = update.rules_config
    if update.rules_format is not None:
        entity.rules_format = update.rules_format
    
    # Increment version on any update
    entity.version += 1
    
    await db.commit()
    await db.refresh(entity)
    
    return EntityResponse(
        id=str(entity.id),
        project_id=str(entity.project_id),
        name=entity.name,
        description=entity.description,
        version=entity.version,
        schema=entity.schema_json,
        rules_config=entity.rules_config,
        rules_format=entity.rules_format,
        created_at=entity.created_at.isoformat(),
        updated_at=entity.updated_at.isoformat(),
    )


@router.delete("/{entity_id}")
@router.put("/{entity_id}", response_model=EntityResponse)
@router.put("/{entity_id}", response_model=EntityResponse)
async def delete_entity(
    project_id: UUID,
    entity_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Delete a wizard entity."""
    stmt = select(WizardEntity).where(
        WizardEntity.id == entity_id,
        WizardEntity.project_id == project_id,
    )
    result = await db.execute(stmt)
    entity = result.scalar_one_or_none()
    
    if not entity:
        raise HTTPException(status_code=404, detail="Entity not found")
    
    await db.delete(entity)
    await db.commit()
